### PR TITLE
Pybind11-ify rclpy_get_node_parameters

### PR DESCRIPTION
**Pull request by sloretz**
Part of #665

### CODE REVIEW

**Comment by sloretz** [comment]
CI (build: `--packages-up-to rclpy` test: `--packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14038/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8859/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11731/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14129/)

File: rclpy/src/rclpy/node.hpp [hunk 1]
```diff
@@ -42,6 +42,20 @@ get_node_name(py::capsule pynode);
  */
 const char *
 get_node_namespace(py::capsule pynode);
+
+/// Get a list of parameters for the current node
+/**
+ * On failure, an exception is raised and NULL is returned if:
```
Comment (hidmic) [inline rclpy/src/rclpy/node.hpp:48]: @sloretz safe to delete?

File: rclpy/src/rclpy/node.cpp [hunk 2]
```diff
@@ -62,4 +69,192 @@ get_node_namespace(py::capsule pynode)
...
+ * \param[out] params_by_node_name A Python dict object to place parsed parameters into.
+ */
+void
+_parse_param_overrides(
+  const rcl_arguments_t * args, py::object pyparameter_cls,
+  py::object pyparameter_type_cls, py::dict pyparams_by_node_name)
+{
+  rcl_params_t * params = nullptr;
+  if (RCL_RET_OK != rcl_arguments_get_param_overrides(args, &params)) {
+    throw RCLError("failed to get parameter overrrides");
```
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:193]: @sloretz typo:  
```suggestion
    throw RCLError("failed to get parameter overrides");
```

File: rclpy/src/rclpy/node.cpp [hunk 3]
```diff
@@ -62,4 +69,192 @@ get_node_namespace(py::capsule pynode)
...
+/**
+ * On failure, an exception is raised and NULL is returned if:
+ *
+ * Raises ValueError if the argument is not a node handle.
+ * Raises RCLError if the parameters file fails to parse
+ *
+ * \param[in] pyparameter_cls The rclpy.parameter.Parameter class object.
+ * \param[in] node_capsule Capsule pointing to the node handle
+ * \return A dict mapping parameter names to rclpy.parameter.Parameter (may be empty).
+ */
```
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:212]: @sloretz nit: isn't this docblock duplicated in the header file?

File: rclpy/src/rclpy/node.hpp [hunk 4]
```diff
@@ -42,6 +42,20 @@ get_node_name(py::capsule pynode);
...
+ *
+ * Raises ValueError if the argument is not a node handle.
+ * Raises RCLError if the parameters file fails to parse
```
Comment (hidmic) [inline rclpy/src/rclpy/node.hpp:51]: @sloretz meta: I think there are quite a few more failure modes.

**Review by hidmic — APPROVED** [review]
LGTM but for some minor comments !

File: rclpy/src/rclpy/node.cpp [hunk 5]
```diff
@@ -62,4 +69,192 @@ get_node_namespace(py::capsule pynode)
...
+    if ('/' != node_name.front()) {
+      node_name.insert(node_name.begin(), '/');
+    }
+    auto pynode_name = py::str(node_name);
+
+    // make a dictionary for the parameters belonging to this specific node name
+    if (!pynode_params.contains(pynode_name)) {
+      pynode_params[pynode_name] = py::dict();
+    }
+    auto parameter_dict = pynode_params[pynode_name].cast<py::dict>();
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/src/rclpy/node.cpp]: Implicit cast in 3ed7cb8311d1fb28bbec294c2dc901392684b31e

File: rclpy/src/rclpy/node.cpp [hunk 6]
```diff
@@ -62,4 +69,192 @@ get_node_namespace(py::capsule pynode)
...
+
+    rcl_node_params_t node_params = params->params[i];
+    for (size_t ii = 0; ii < node_params.num_params; ++ii) {
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/src/rclpy/node.cpp]: `j` instead of `ii` in 18197c17aa74dee3683084900f1413648dfc42c3

File: rclpy/src/rclpy/node.hpp (hunk 1 above)
Comment (sloretz) [reply rclpy/src/rclpy/node.hpp]: Deleted this text in 33c40a34f62ff30c6490a03a31222d40250e0fc6

File: rclpy/src/rclpy/node.cpp (hunk 3 above)
Comment (sloretz) [reply rclpy/src/rclpy/node.cpp]: Removed duplicate docblock in 33c40a34f62ff30c6490a03a31222d40250e0fc6

File: rclpy/src/rclpy/node.hpp (hunk 4 above)
Comment (sloretz) [reply rclpy/src/rclpy/node.hpp]: Documented more exceptions in 33c40a34f62ff30c6490a03a31222d40250e0fc6

File: rclpy/src/rclpy/node.cpp (hunk 2 above)
Comment (sloretz) [reply rclpy/src/rclpy/node.cpp]: overrrides -> overrides in be8fa76bb89941318be2fe641d3b7b30cea1c37e

**Comment by sloretz** [comment]
CI with feedback addressed (build: `--packages-up-to rclpy` test: `--packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14050/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8871/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11743/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14141/)

**Comment by sloretz** [comment]
CI LGTM, @hidmic may I ask you to double check that the changes address your feedback?

**Review by hidmic — APPROVED** [review]
LGTM !

**Pull request merged**